reader.go

```diff
@@ -169,43 +225,60 @@ func (d *decoder) parseConfig(e entry) (cfg image.Config, err error) {
 
 func (d *decoder) setupBMP(e entry, data []byte) ([]byte, []byte, int, error) {
 	// Ico files are made up of a XOR mask and an AND mask
-	// The XOR mask is the image itself, while the AND mask is a 1 bit-per-pixel alpha channel.
+	// The XOR mask is the image itself, while the AND mask is a 1 bit-per-pixel alpha channel (transparent or opaque).
 	// setupBMP returns the image as a BMP format byte array, and the mask as a (1bpp) pixel array
 
 	// calculate image sizes
 	// See wikipedia en.wikipedia.org/wiki/BMP_file_format
-	var imageSize, maskSize int
-	if int(e.Size) < len(data) {
-		imageSize = int(e.Size)
-	} else {
-		imageSize = len(data)
-	}
+	var imageSize, maskSize uint32
+	imageSize = uint32(len(data))
 	if e.Bits != 32 {
-		rowSize := (1 * (int(e.Width) + 31) / 32) * 4
-		maskSize = rowSize * int(e.Height)
+		rowSize := (1 * (uint32(e.Width) + 31) / 32) * 4
+		maskSize = rowSize * uint32(e.Height)
+		if maskSize > imageSize {
+			return nil, nil, 0, FormatError("masksize exceeds image size")
+		}
 		imageSize -= maskSize
 	}
 
-	img := make([]byte, 14+imageSize)
-	mask := make([]byte, maskSize)
+	if len(data) < int(imageSize) {
+		return nil, nil, 0, FormatError(
+			fmt.Sprintf("datalen %d smaller than imageSize %d", len(data), imageSize))
+	}
+
+	if 14+imageSize < 50 {
+		return nil, nil, 0, FormatError(fmt.Sprintf("imagesize too small: %d", imageSize))
+	}
+
+	img, err := d.allocMemory(14 + imageSize)
+	if err != nil {
+		return nil, nil, 0, fmt.Errorf("failed to allocate image buffer: %w", err)
+	}
+
+	mask, err := d.allocMemory(maskSize)
+	if err != nil {
+		return nil, nil, 0, fmt.Errorf("failed to allocate mask buffer: %w", err)
+	}
 
-	var n int
+	var n uint32
 	// Read in image
-	n = copy(img[14:], data[:imageSize])
+	n = uint32(copy(img[14:], data[:imageSize]))
 	if n != imageSize {
 		return nil, nil, 0, FormatError(fmt.Sprintf("only %d of %d bytes read.", n, imageSize))
 	}
 	// Read in mask
-	n = copy(mask, data[imageSize:])
+	n = uint32(copy(mask, data[imageSize:]))
 	if n != maskSize {
 		return nil, nil, 0, FormatError(fmt.Sprintf("only %d of %d bytes read.", n, maskSize))
 	}
 
-	var dibSize, w, h uint32
-	binary.Read(bytes.NewReader(img[14:14+4]), binary.LittleEndian, &dibSize)
-	binary.Read(bytes.NewReader(img[14+4:14+8]), binary.LittleEndian, &w)
-	binary.Read(bytes.NewReader(img[14+8:14+12]), binary.LittleEndian, &h)
+	// following slices will not panic due to image size check above
 
+	dibSize := binary.LittleEndian.Uint32(img[14 : 14+4])
+	w := binary.LittleEndian.Uint32(img[14+4 : 14+8])
+	h := binary.LittleEndian.Uint32(img[14+8 : 14+12])
+
+	// what case is this handling?
 	if h > w {
```

Comment:
noting that I looked at this and the original repo had a single commit. Unsure on what case this is handling